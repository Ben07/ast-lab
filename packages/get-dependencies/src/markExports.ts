import { ALL_EXPORTS } from './constants';
import { ModuleExported, Exports, Exported } from './types';

function markExports(marked: Exports, source: string, markExports: ModuleExported) {
  const markedMod = marked.get(source);
  const isLimitedExport = markExports instanceof Set;
  if (!markedMod) {
    marked.set(
      source,
      isLimitedExport ? new Set(Array.from(markExports as Set<Exported>)): markExports
    );
  } else if (markedMod !== ALL_EXPORTS) {
    if (isLimitedExport) {
      (markExports as Set<Exported>).forEach(v => markedMod.add(v));
    } else if (markExports === ALL_EXPORTS) {
      marked.set(source, ALL_EXPORTS);
    }
  }
  return marked;
}

export default markExports;
