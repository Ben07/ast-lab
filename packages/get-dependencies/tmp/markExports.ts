import { ALL_EXPORTS } from './constants';
import { ModuleExported, Exports, Exported } from './types';

function markExports(marked: Exports, source: string, markExports: ModuleExported) {
  const markedMod = marked.get(source);
  const isLimitedExport = markExports instanceof Set;
  if (!markedMod) {
    return marked.set(
      source,
      isLimitedExport ? new Set(Array.from(markExports as Set<Exported>)): markExports
    );
  } else if (markedMod !== ALL_EXPORTS) {
    if (markExports === ALL_EXPORTS) {
      return marked.set(source, ALL_EXPORTS);
    } else if (isLimitedExport) {
      const updatedExports = Array.from(markedMod);
      (markExports as Set<Exported>).forEach(v => {
        if (!markedMod.has(v)) {
          updatedExports.push(v);
        }
      });
      // create new Set for easier comparing
      if (updatedExports.length !== markedMod.size) {
        return marked.set(source, new Set(updatedExports));
      }
    } 
  }
  return marked;
}

export default markExports;
