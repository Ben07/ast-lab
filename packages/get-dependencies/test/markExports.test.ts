// @ts-ignore
import { enumerateArgsTestFunction, configArgs } from 'lazy-jest';
import markExports from '../src/markExports';

describe('markExports()', () => {
  enumerateArgsTestFunction(
    markExports,
    configArgs()
      .arg('marked', [new Map()])
      .arg('source', ['a'])
      .arg('markExports', [
        null,
      ]),
    'no exports should not mark'
  );
  enumerateArgsTestFunction(
    markExports,
    configArgs()
      .arg('marked', [new Map()])
      .arg('source', ['a'])
      .arg('markExports', [
        true,
        null,
        new Set(['a1', 'a2', 'a3'])
      ]),
    'all exports should not be overwritten'
  );
  enumerateArgsTestFunction(
    markExports,
    configArgs()
      .arg('marked', [new Map()])
      .arg('source', ['a'])
      .arg('markExports', [
        new Set(['a1', 'a3']),
        new Set(['a2', 'a3']),
      ]),
    'exports should append and dedupe'
  );
  enumerateArgsTestFunction(
    markExports,
    configArgs()
      .arg('marked', [new Map()])
      .arg('source', ['a'])
      .arg('markExports', [
        null,
        new Set(['a1',]),
        new Set(['a2', 'a3']),
      ]),
    'no exports should be overwritten by limited exports'
  );
  enumerateArgsTestFunction(
    markExports,
    configArgs()
      .arg('marked', [new Map()])
      .arg('source', ['a'])
      .arg('markExports', [
        new Set(['a1', 'a2', 'a3']),
        true
      ]),
    'limited exports should be overwritten by all exports'
  );
});
