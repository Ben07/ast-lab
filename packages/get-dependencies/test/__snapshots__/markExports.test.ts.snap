// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`markExports() all exports should not be overwritten - (["marked", "source", "markExports"]) - [{},"a",{}] 1`] = `
Map {
  "a" => true,
}
`;

exports[`markExports() all exports should not be overwritten - (["marked", "source", "markExports"]) - [{},"a",null] 1`] = `
Map {
  "a" => true,
}
`;

exports[`markExports() all exports should not be overwritten - (["marked", "source", "markExports"]) - [{},"a",true] 1`] = `
Map {
  "a" => true,
}
`;

exports[`markExports() exports should append and dedupe - (["marked", "source", "markExports"]) - [{},"a",{}] 1`] = `
Map {
  "a" => Set {
    "a1",
    "a3",
  },
}
`;

exports[`markExports() exports should append and dedupe - (["marked", "source", "markExports"]) - [{},"a",{}] 2`] = `
Map {
  "a" => Set {
    "a1",
    "a3",
    "a2",
  },
}
`;

exports[`markExports() limited exports should be overwritten by all exports - (["marked", "source", "markExports"]) - [{},"a",{}] 1`] = `
Map {
  "a" => Set {
    "a1",
    "a2",
    "a3",
  },
}
`;

exports[`markExports() limited exports should be overwritten by all exports - (["marked", "source", "markExports"]) - [{},"a",true] 1`] = `
Map {
  "a" => true,
}
`;

exports[`markExports() no exports should be overwritten by limited exports - (["marked", "source", "markExports"]) - [{},"a",{}] 1`] = `
Map {
  "a" => Set {
    "a1",
  },
}
`;

exports[`markExports() no exports should be overwritten by limited exports - (["marked", "source", "markExports"]) - [{},"a",{}] 2`] = `
Map {
  "a" => Set {
    "a1",
    "a2",
    "a3",
  },
}
`;

exports[`markExports() no exports should be overwritten by limited exports - (["marked", "source", "markExports"]) - [{},"a",null] 1`] = `
Map {
  "a" => null,
}
`;

exports[`markExports() no exports should not mark - (["marked", "source", "markExports"]) - [{},"a",null] 1`] = `
Map {
  "a" => null,
}
`;
